sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveTooManyLines.cs

```diff
@@ -0,0 +1,117 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class MethodsShouldNotHaveTooManyLines : ParameterLoadingDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S138";
+        private const string MessageFormat = "This {0} has {1} lines, which is greater than the {2} lines authorized. Split it into smaller methods.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private const int DefaultMaxMethodLines = 80;
+        [RuleParameter("max", PropertyType.Integer, "Maximum authorized lines of code in a method",
+            DefaultMaxMethodLines)]
+
+        public int Max { get; set; } = DefaultMaxMethodLines;
+        
+        protected override void Initialize(ParameterLoadingAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    if (Max < 2)
```

Comment:
This is what other paranetrized rules do. I haven't found a way to test that.

Reply:
Which parameterised rules check the parameter and throw? The only example I could find was _CheckFileLicense_ - all of the others I looked at didn't seem to validate the supplied parameter at all.

I think validation is a good idea, but is throwing the best way to report an error? Will the user see the error message, or will they just get a _Analyzer X has thrown an exception_ gold bar in the Error List?
(I'm not sure how I would report the invalid parameter, but I'm concerned that throwing makes it looks like the analyzer is buggy).

Reply:
That's the one I found. The message I get in test is actually pretty good:
`error AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.MethodsShouldNotHaveTooManyLines' threw an exception of type 'System.InvalidOperationException' with message 'Invalid maximum number of lines: -1'.`

I prefer an exception (even if it's a bit ugly) instead of silent failure (or silent reset to default value). Not sure what are other options of reporting error, I'll investigate (but that's outside of scope of this PR).

Reply:
Sounds good. I'm not sure there are many other options.

It isn't clear from the error message that the problem is with a parameter supplied to the rule (particularly since users might not know that rules can parameterised). Something like _Invalid rule parameter: maximum number of lines = -1. Must be at least 2._

Would throwing _ArgumentOutOfRangeException_ also help? We should consider defining our own custom exception type (outside the scope of this commit).